core/server/master/src/main/java/alluxio/master/file/DefaultFileSystemMaster.java

```diff
@@ -1145,6 +1162,113 @@ public List<FileInfo> listStatus(AlluxioURI path, ListStatusContext context)
     return fileInfos;
   }
 
+  /**
+   * If performing a partial listing, from an offset, this checks to ensure the offset
+   * exists within the starting path.
+   * @param path the listing path
+   * @param context the listing context
+   * @return the components of the path to the offset
+   */
+  private @Nullable List<String> checkPartialListingOffset(
+      AlluxioURI path, ListStatusContext context)
+      throws FileDoesNotExistException, InvalidPathException {
+    List<String> partialPathNames = null; // null if we are not using a partial listing
+    if (context.isPartialListing() && context.getOptions().getOffset() != 0) {
+      try {
+        // See if the inode from where to start the listing exists.
+        partialPathNames = mInodeTree.getPathInodeNames(context.getOptions().getOffset());
+      } catch (FileDoesNotExistException e) {
+        throw new FileDoesNotExistException(
+            ExceptionMessage.INODE_NOT_FOUND_PARTIAL_LISTING.getMessage(e.getMessage()),
+            e.getCause());
+      }
+      String[] pathComponents = PathUtils.getPathComponents(path.getPath());
+      // the offset path must be at least as long as the starting path
+      if (partialPathNames.size() < pathComponents.length) {
+        throw new FileDoesNotExistException(ExceptionMessage.INODE_NOT_FOUND_PARTIAL_LISTING
+            .getMessage(path));
+      }
+      for (int i = 0; i < pathComponents.length; i++) {
+        if (!partialPathNames.get(i).equals(pathComponents[i])) {
+          throw new FileDoesNotExistException(ExceptionMessage.INODE_NOT_FOUND_PARTIAL_LISTING
+              .getMessage(path));
+        }
+      }
+    }
+    return partialPathNames;
+  }
+
+  /**
+   * If this is a partial listing, this will compute the path components from where to start
+   * the listing from that come after the root listing path.
+   *
+   * @param context the context of the operation
+   * @param pathNames the full path from where the partial listing is expected to start,
+   *                  null if this is the first listing
+   * @param rootPath the locked root path of the listing
+   * @return the path components after the root path from where to start the partial listing,
+   * or null if the listing should start from the beginning of the root path
+   */
+  private @Nullable List<String> computePartialListingPaths(
+      ListStatusContext context,
+      @Nullable List<String> pathNames, LockedInodePath rootPath)
+      throws InvalidPathException {
+    if (pathNames == null) {
+      // use the start after option, since this is the first listing
+      if (!context.getOptions().getStartAfter().isEmpty()) {
+        return Arrays.stream(PathUtils.getPathComponents(
+            context.getOptions().getStartAfter())).skip(1).collect(Collectors.toList());
```

Comment:
`Collectors.toList()` will create a default size `ArrayList` which will go through resizing and all. If possible to determine or even guess the size upfront, doing that manually may save a bit of memory.

Reply:
Ok I can change this. It is surprising to me though, here the stream is sized, so I would expect it to use this to correctly allocate the array list. I guess java streams aren't up to the standards of other languages.

Reply:
Yea that's unexpectedly dumb. I suspect it is different across JVM implementations, like if you use openjdk nobody cares. However we do have to deal with the dumbest of them all.